src/recipe_parser.zig

```diff
@@ -0,0 +1,264 @@
+				std.log.info("{s}: {s}", .{item.key_ptr.*, item.value_ptr.*});
+			}
+		}
+
+		var iter = items.iterator();
+		loop: while(iter.next()) |item| {
+			for(tags.items) |tag| {
+				if(!item.hasTag(tag) and !(item.block() != null and (Block{.typ = item.block().?, .data = 0}).hasTag(tag))) {
+					continue :loop;
+				}
+			}
+			if(matchWithKeys(allocator, item.id(), pattern.items, keys)) |newKeys| {
+				itemPairs.append(.{
+					.item = .{
+						.item = .{.baseItem = item.*},
+						.amount = amount,
+					},
+					.keys = newKeys,
+				});
+			} else if(id.len == 0) {
+				itemPairs.append(.{
+					.item = .{
+						.item = .{.baseItem = item.*},
+						.amount = amount,
+					},
+					.keys = keys.clone() catch unreachable,
+				});
+			}
+		}
+	}
+	return itemPairs;
+}
+
+fn generateItemCombos(allocator: NeverFailingAllocator, recipe: []ZonElement) !main.List([]ItemStack) {
+	var remainingItems = recipe;
+	var emptyKeys: std.StringHashMap([]const u8) = .init(allocator.allocator);
+	defer emptyKeys.deinit();
+	const startingParsedItems = try parseRecipeItem(allocator, remainingItems[0], &emptyKeys);
+	defer startingParsedItems.deinit();
+	var inputCombos: main.List([]ItemStack) = .initCapacity(allocator, startingParsedItems.items.len);
+	errdefer {
+		for(inputCombos.items) |combo| {
+			allocator.free(combo);
+		}
+		inputCombos.deinit();
+	}
+	var keyList: main.List(std.StringHashMap([]const u8)) = .initCapacity(allocator, startingParsedItems.items.len);
+	defer {
+		for(keyList.items) |*keys| {
+			keys.deinit();
+		}
+		keyList.deinit();
+	}
+	for(startingParsedItems.items) |item| {
+		const inputs = allocator.alloc(ItemStack, recipe.len);
+		inputs[0] = item.item;
+		inputCombos.append(inputs);
+		keyList.append(item.keys);
+	}
+	while(remainingItems.len > 1) {
+		remainingItems = remainingItems[1..];
+		const startIndex = inputCombos.items[0].len - remainingItems.len;
+		var newKeyList: main.List(std.StringHashMap([]const u8)) = .init(allocator);
+		var newInputCombos: main.List([]ItemStack) = .init(allocator);
+
+		for(keyList.items, inputCombos.items) |*keys, inputs| {
+			const parsedItems = try parseRecipeItem(allocator, remainingItems[0], keys);
+			defer parsedItems.deinit();
+			for(parsedItems.items) |item| {
+				const newInputs = allocator.dupe(ItemStack, inputs);
+				newInputs[startIndex] = item.item;
+				newInputCombos.append(newInputs);
+				newKeyList.append(item.keys);
+			}
+		}
+
+		for(keyList.items) |*keys| {
+			keys.deinit();
+		}
+		keyList.deinit();
+		keyList = newKeyList;
+		for(inputCombos.items) |combo| {
+			allocator.free(combo);
+		}
+		inputCombos.deinit();
+		inputCombos = newInputCombos;
+	}
+	return inputCombos;
+}
+
+pub fn parseRecipe(allocator: NeverFailingAllocator, zon: ZonElement, list: *main.List(Recipe)) !void {
+	const inputs = zon.getChild("inputs").toSlice();
+	const recipeItems = allocator.alloc(ZonElement, inputs.len + 1);
+	@memmove(recipeItems[0..inputs.len], inputs);
```

Comment:
If newly allocated memory aliases with another slice then that would be a sign of a big problem
```suggestion
	@memcpy(recipeItems[0..inputs.len], inputs);
```